File: Python/SubtreeOfAnotherTree/SubtreeOfAnotherTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
# (119ms)
    def isSubtree(self, s, t):
        """
        :type s: TreeNode
        :type t: TreeNode
        :rtype: bool
        """
        def traverse(root, now):
            if not root:
                now.append("$")
                return
            now.append(`root.val`)
            traverse(root.left, now)
            traverse(root.right, now)
        s_list, t_list = [], []
        traverse(s, s_list)
        traverse(t, t_list)
        s_str, t_str= "," + ",".join(s_list), "," + ",".join(t_list)
        return t_str in s_str

File: Java/SubtreeOfAnotherTree/SubtreeOfAnotherTree.java
import com.phoenix.Util.Tree;
import com.phoenix.Util.TreeNode;

/**
 * (27ms)
 * 这个方法是遍历s树中的每一个点，然后调用子函数isSame来判断以当前点为根节点的子树是否与树t相同。
 * Created by Phoenix-Z on 2017/5/8.
 */
public class SubtreeOfAnotherTree {

    public boolean isSubtree(TreeNode s, TreeNode t) {
        if(s == null)
            return false;
        return isSame(s, t) || isSubtree(s.left, t) || isSubtree(s.right, t);
    }

    private boolean isSame(TreeNode s, TreeNode t) {
        if(s == null && t == null)
            return true;
        if(s == null || t == null)
            return false;
        if(s.val != t.val)
            return false;
        return isSame(s.left, t.left) && isSame(s.right, t.right);
    }

    public static void main(String[] args) {
        TreeNode s = Tree.getTree(new Integer[] {3, 4, 5, 1, 2, null, null, null, null, 0});
        TreeNode t = Tree.getTree(new Integer[] {4, 1, 2});
        SubtreeOfAnotherTree test = new SubtreeOfAnotherTree();
        System.out.println(test.isSubtree(s, t));
    }

}
